numbat/src/command.rs

```diff
@@ -0,0 +1,436 @@
+    word_boundaries: &[(u32, u32)],
+    word_index: usize,
+    code_source_id: usize,
+) -> Span {
+    let start = word_boundaries[word_index].0;
+    let end = word_boundaries.last().unwrap().1;
+    span_from_boundary((start, end), code_source_id)
+}
+
+fn span_from_boundary((start, end): (u32, u32), code_source_id: usize) -> Span {
+    Span {
+        start: SourceCodePositition {
+            byte: start,
+            line: 1,
+            position: start,
+        },
+        end: SourceCodePositition {
+            byte: end,
+            line: 1,
+            position: end,
+        },
+        code_source_id,
+    }
+}
+
+macro_rules! handle_arg_count {
+    (count = 0, $words:expr, $word_boundaries:expr, $code_source_id:expr, "help", $if_ok:expr $(,)?) => {{
+        handle_arg_count!(
+            count = 0,
+            $words,
+            $word_boundaries,
+            $code_source_id,
+            "help",
+            Command::Help,
+            "`help` takes 0 arguments; use `info <item>` for information about an item",
+        )
+    }};
+    (count = 0, $words:expr, $word_boundaries:expr, $code_source_id:expr, $command:literal, $if_ok:expr $(,)?) => {{
+        handle_arg_count!(
+            count = 0,
+            $words,
+            $word_boundaries,
+            $code_source_id,
+            $command,
+            $if_ok,
+            concat!("`", $command, "` takes 0 arguments")
+        )
+    }};
+    (count = 0, $words:expr, $word_boundaries:expr, $code_source_id:expr, $command:literal, $if_ok:expr, $err_msg:expr $(,)?) => {{
+        if $words.next().is_some() {
+            return Some(Err(ParseError {
+                kind: ParseErrorKind::InvalidCommand($err_msg),
+                span: span_through_end($word_boundaries, 1, $code_source_id),
+            }));
+        }
+
+        $if_ok
+    }};
+}
+
+/// this function primarily exists for testing, to not have to provide a resolver
+///
+/// for an actual build, a resolver is necessary for error reporting
+fn parse_command_inner<'a>(
+    input: &'a str,
+    resolver: Option<&mut Resolver>,
+) -> Option<Result<Command<'a>, ParseError>> {
+    let word_boundaries = get_word_boundaries(input);
+
+    let code_source_id = match resolver {
+        Some(resolver) => resolver.add_code_source(CodeSource::Text, input),
+        None => 0,
+    };
+
+    let mut words = input.split_whitespace();
+    let Some(command_str) = words.next() else {
+        // should never hit this branch in practice because all-whitespace inputs are
+        // skipped over
+
+        return Some(Err(ParseError {
+            kind: ParseErrorKind::InvalidCommand("invalid empty command"),
+            span: span_from_boundary((0, u32::try_from(input.len()).unwrap()), code_source_id),
+        }));
+    };
+
+    let command = match command_str {
+        "help" => handle_arg_count!(
+            count = 0,
+            words,
+            &word_boundaries,
+            code_source_id,
```

Comment:
I wonder if you could get rid of passing all these arguments every time if you implement the command parser as a `struct` with member variables?